Add kluster spec migration

databus23 opened the pull request:
This is an attempt provide a fix for #97.
Specifically this PR provides a lightweight way of automatically migrating kluster specs to newer versions (e.g. add/rename fields with auto discovered values etc).
This is **not** intended to facilitate upgrades of klusters (`helm upgrades`) but update our internal representation of a kluster.

A little explanation:
* The ordered list of migration steps is kept in a go slice in `pkg/migration/register.go`. The versioning is based on simply enumerating the migrations steps (4 migrations steps means the latest version is 4) 
* The spec/schema version of a Kluster is kept in `Status.SpecVersion`. New Klusters are created with the latest Version of the apiserver.
* A migration step is defined by a reasonable simple function signature (see `migration.Migration`)
* During operator startup before the controllers are started all existing klusters pending migrations are upgraded. I did it explicitly this way to avoid loosing the raw Data from a previous spec (e.g. a field was removed). Because as soon as a controller updates a kluster the internal json representation is completely replaced.

Open questions:
1. When upgrading a kluster fails it is not retried until the operator is restarted. Do we want to have a controller type logic that reconciles this continuously? If so we need to also think if we want to have controllers ignore klusters that are pending migrations. Otherwise we risk having them operate on incomplete data and destroy old information that would maybe be required for completing the migration.
2. Should be maybe call it `SchemaVersion` instead of `SpecVersion` to avoid confusion?

SchwarzM commented:
Hi,

i am currently missing code that forbids a version 3 operator to handle version 4 klusters.
In general I think during a rolling update we will have version 3 and version 4 operators in parallel.
If I would inject a version 1 kluster (with kubernikusctl) both would try to migrate it.

Kind regards
M. Schwarz

BugRoger commented:
As discussed:

- Migrations run before the operators are started
- If a migration fails, the kluster will change phase to `disabled`
- Controllers are supposed to ignore `disabled` klusters
- Add a Migration controller that periodically tries to migrate `disabled` klusters and emit events
- Change phase updates to `patch` in order to keep outdated spec values
- Please also add a little bit of documentation to `docs/development/migrations.md` explaining how this works and what the expectations are

databus23 commented:
On second thought introducing a `Disabled` phase is not a good idea as we will have the same problems we had with the `Error` phase: We don't know which phase the Kluster was in before it got disabled. I think we need to express this as some kind of `Condition`. Maybe as simple as a boolean `Status.Disabled`

databus23 commented:
This is ready for another review round. I added a small amount of documentation explaining the basic.
Here are the basics:
* I added a *Condition* `Status.MigrationsPending`, that is updated before controllers start.
* Controllers that modify the Kluster are supposed to check `Disabled()` on a `Kluster` resource and exit early. The idea is that in the future other conditions might disable a kluster and controller don't need to know the specifics just that they need to ignore this Kluster
* The migration controller reconciles `Status.MigrationStatus` for all klusters continuously (tries to migrate klusters that have migrations pending).

databus23 commented:
@BugRoger Care ro review? I consider this ready fo review.

BugRoger commented on docs/development/migrations.md:
````diff
@@ -0,0 +1,22 @@
+```
+type Migration func(klusterRaw []byte, kluster *v1.Kluster, client kubernetes.Interface) (err error)
+```
+* `klusterRaw` contains the current raw JSON data of the Kluster, it can be used to access old data not exposed by the `v1.Kluster` type anymore.
````
I have a vague recollection that we discussed about this, but why is `klusterRaw` a `json` blob? Wouldn't it be easier to access if it were a deep-copy of `kluster`?

databus23 replied:
The problem is that the migration runs in the new codebase with the new `CRD` model. That means if you removed a field (or renamed, changed the structure) you loose the old data. So loading it into a `Kluster` is not an option.

BugRoger commented on docs/development/migrations.md:
```diff
@@ -0,0 +1,22 @@
+Klusters with pending migrations are ignored by all controllers that modify the internal Kluster representation (e.g. update the CRD).
+`Status.MigrationsPending` is updated during operator startup before any controller is running and continuously kept in sync by the migration controller.
+
+**Controllers should call the `Disabled()` method on the `v1.Kluster` and exit early in reconciliation functions in case a Kluster is disabled.**
```
Sounds like we have to add this into the other existing controller loops once this is merged.

BugRoger replied:
Ah, nvm.

BugRoger commented on pkg/controller/migration/controller.go:
```diff
@@ -0,0 +1,65 @@
+
+func (mr *MigrationReconciler) Reconcile(kluster *v1.Kluster) (bool, error) {
+
+	//We only care about klustes with pending migrations
```
kluster

databus23 replied:
Fixed, thanks.

BugRoger commented on pkg/controller/migration/controller.go:
```diff
@@ -0,0 +1,65 @@
+func (mr *MigrationReconciler) Reconcile(kluster *v1.Kluster) (bool, error) {
+
+	//We only care about klustes with pending migrations
+	if !migration.MigrationsPending(kluster) {
```
These too loops are a not obvious. Double negation and noop in the called function. Feels like it could be unraveled.

databus23 replied:
I'm not sure I follow. Care to elaborate?

BugRoger commented on pkg/controller/operator.go:
```diff
@@ -185,8 +188,11 @@ func (o *KubernikusOperator) Run(stopCh <-chan struct{}, wg *sync.WaitGroup) {
-	o.Logger.Log("msg", "Cache primed. Ready for Action!")
-
+	o.Logger.Log("msg", "Cache primed")
+	if _, found := o.Config.Kubernikus.Controllers["migration"]; found {
```
Is this really necessary here? Wouldn't the immediate first reconciliation be equivalent?

databus23 replied:
The idea was to have the klusters migrated before any controllers are running to avoid running controllers that expect the new model on an old model that hasn't been migrated yet.
Also to not accidentally modify models that haven't been migrated yet (so that we don't loose the old data we might need for the migration) we need to make sure they are updated to be `Disabled()` before other controllers run.

BugRoger replied:
Ah, ok. Good. Carry on... :)

BugRoger commented on pkg/migration/migration.go:
```diff
@@ -0,0 +1,90 @@
+	var err error
+
+	//TODO: Don't import fake pkg outside of test code
+	if _, ok := kubernikus_client.(*kubernikusfake.Clientset); !ok {
```
I suppose the FakeClient doesn't mock the `RESTClient()`?

databus23 replied:
Spot on.

BugRoger commented on pkg/migration/migration.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+//AddMigration appends a migration to the list
+func (r *Registry) AddMigration(m Migration) {
```
Hm. Coming from Rails where the migrations are ordered by a unique date prefix, I was confused how this works here. I think, this `Registry` assumes that the migrations are always added at the end of the `register` function. And that there is no change in order or deletions.

I think, it's cool to keep it simple. But we should probably add a comment into the register call to not change it. Especially to not delete previously commited migrations.

databus23 replied:
You are absolutely right with your assessments. I'll add the comment and would not invest more time at this point to make it more like a full-blown migration system (e.g. as in rails).

BugRoger replied:
Cool!

BugRoger commented on pkg/migration/register.go:
```diff
@@ -0,0 +1,7 @@
+package migration
+
+func init() {
```
Add safeguard comment to not mess around with existing migrations.

databus23 replied:
done
